File: identity_iota_core/src/rebased/iota/move_calls/asset/transfer.rs
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use identity_iota_interaction::types::base_types::IotaAddress;
use identity_iota_interaction::types::base_types::ObjectID;
use identity_iota_interaction::types::base_types::ObjectRef;
use identity_iota_interaction::types::base_types::SequenceNumber;
use identity_iota_interaction::types::programmable_transaction_builder::ProgrammableTransactionBuilder;
use identity_iota_interaction::types::transaction::Command;
use identity_iota_interaction::types::transaction::ObjectArg;
use identity_iota_interaction::types::transaction::ProgrammableTransaction;
use identity_iota_interaction::types::TypeTag;
use move_core_types::ident_str;

use identity_iota_interaction::MoveType;
use crate::rebased::Error;

pub(crate) fn transfer<T: MoveType>(
  asset: ObjectRef,
  recipient: IotaAddress,
  package: ObjectID,
) -> Result<ProgrammableTransaction, Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let asset = ptb
    .obj(ObjectArg::ImmOrOwnedObject(asset))
    .map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let recipient = ptb.pure(recipient).map_err(|e| Error::InvalidArgument(e.to_string()))?;

  ptb.command(Command::move_call(
    package,
    ident_str!("asset").into(),
    ident_str!("transfer").into(),
    vec![T::move_type(package)],
    vec![asset, recipient],
  ));

  Ok(ptb.finish())
}

fn make_tx(
  proposal: (ObjectID, SequenceNumber),
  cap: ObjectRef,
  asset: ObjectRef,
  asset_type_param: TypeTag,
  package: ObjectID,
  function_name: &'static str,
) -> Result<ProgrammableTransaction, Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let proposal = ptb
    .obj(ObjectArg::SharedObject {
      id: proposal.0,
      initial_shared_version: proposal.1,
      mutable: true,
    })
    .map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let cap = ptb
    .obj(ObjectArg::ImmOrOwnedObject(cap))
    .map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let asset = ptb
    .obj(ObjectArg::Receiving(asset))
    .map_err(|e| Error::InvalidArgument(e.to_string()))?;

  ptb.command(Command::move_call(
    package,
    ident_str!("asset").into(),
    ident_str!(function_name).into(),
    vec![asset_type_param],
    vec![proposal, cap, asset],
  ));

  Ok(ptb.finish())
}

pub(crate) fn accept_proposal(
  proposal: (ObjectID, SequenceNumber),
  recipient_cap: ObjectRef,
  asset: ObjectRef,
  asset_type_param: TypeTag,
  package: ObjectID,
) -> Result<ProgrammableTransaction, Error> {
  make_tx(proposal, recipient_cap, asset, asset_type_param, package, "accept")
}

pub(crate) fn conclude_or_cancel(
  proposal: (ObjectID, SequenceNumber),
  sender_cap: ObjectRef,
  asset: ObjectRef,
  asset_type_param: TypeTag,
  package: ObjectID,
) -> Result<ProgrammableTransaction, Error> {
  make_tx(
    proposal,
    sender_cap,
    asset,
    asset_type_param,
    package,
    "conclude_or_cancel",
  )
}

File: bindings/wasm/iota_interaction_ts/lib/move_calls/asset/transfer.ts
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { SharedObjectRef } from "@iota/iota-sdk/dist/cjs/bcs/types";
import { ObjectRef, Transaction } from "@iota/iota-sdk/transactions";

export function transfer(
    asset: ObjectRef,
    assetType: string,
    recipient: string,
    packageId: string,
): Promise<Uint8Array> {
    const tx = new Transaction();
    const assetArg = tx.objectRef(asset);
    const recipientArg = tx.pure.address(recipient);

    tx.moveCall({
        target: `${packageId}::asset::transfer`,
        typeArguments: [assetType],
        arguments: [assetArg, recipientArg],
    });

    return tx.build();
}

function makeTx(
    proposal: SharedObjectRef,
    cap: ObjectRef,
    asset: ObjectRef,
    assetType: string,
    packageId: string,
    functionName: string,
): Promise<Uint8Array> {
    const tx = new Transaction();
    const proposalArg = tx.sharedObjectRef(proposal);
    const capArg = tx.objectRef(cap);
    const assetArg = tx.objectRef(asset);

    tx.moveCall({
        target: `${packageId}::asset::${functionName}`,
        typeArguments: [assetType],
        arguments: [proposalArg, capArg, assetArg],
    });

    return tx.build();
}

export function acceptProposal(
    proposal: SharedObjectRef,
    recipientCap: ObjectRef,
    asset: ObjectRef,
    assetType: string,
    packageId: string,
): Promise<Uint8Array> {
    return makeTx(
        proposal,
        recipientCap,
        asset,
        assetType,
        packageId,
        "accept",
    );
}

export function concludeOrCancel(
    proposal: SharedObjectRef,
    senderCap: ObjectRef,
    asset: ObjectRef,
    assetType: string,
    packageId: string,
): Promise<Uint8Array> {
    return makeTx(
        proposal,
        senderCap,
        asset,
        assetType,
        packageId,
        "conclude_or_cancel",
    );
}
